src/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,208 @@
+
+
+#include <emscripten/emscripten.h>
+#include <emscripten/val.h>
+#include <emscripten/bind.h>
+
+#include <string>
+#include <iostream>
+
+#include "ffi/ir_stream/decoding_methods.hpp"
+#include "spdlog_with_specializations.hpp"
+#include "TimestampPattern.hpp"
+#include "streaming_compression/zstd/Decompressor.hpp"
+#include "ir/LogEventDeserializer.hpp"
+#include "ffi/encoding_methods.hpp"
+
+#ifdef __cplusplus
+#define EXTERN extern "C"
+#else
+#define EXTERN
+#endif
+
+// FIXME: use macro
+enum class LOG_VERBOSITY
+{
+    NONE = 0,
+    TRACE,
+    DEBUG,
+    INFO,
+    WARN,
+    ERROR,
+    FATAL
+};
+
+const char *VERBOSITIES[] = {
+    nullptr,
+    "TRACE",
+    "DEBUG",
+    "INFO",
+    "WARN",
+    "ERROR",
+    "FATAL",
+};
+
+EM_JS(void, insert_row, (emscripten::EM_VAL results_handle, const char *message, int64_t timestamp, int verbosity), {
+    const results = Emval.toValue(results_handle);
+    results.push([ UTF8ToString(message), timestamp, verbosity ]);
+});
+
+class ClpIrV1Decoder
```

Comment:
hmm... This class not only deserializes but also decodes / formats the log events. On the other hand, the Log Viewer has more emphasis on decoding (formatting) log events which is why we named the deserialzing + decoding entity as `Decoder`s: https://github.com/y-scope/yscope-log-viewer/blob/main/new-log-viewer/src/typings/decoders.ts

@kirkrodrigues My understanding is that in our Decoder classes, the `buildIdx` method is for deserializations and the `decode` method is for decoding / formatting. Do you see if we should separate the decoders and deserializers in the Log Viewer? If not, shall we rename the classes / method in the Log Viewer so we can stay consistent across all projects?

Reply:
* Yes, `buildIdx` deserializes and `decode` decodes and formats. (Perhaps we should rename `buildIdx` to `deserializeAndBuffer`?)
* If we separate the deserializer and decoder, I guess the deserializer would deserialize and buffer, and the decoder would take the buffer as input in order to decode? I think that could work and it may be more future proof (as well as leading to smaller and more intuitive classes). Looking at LogFileManager in the log viewer, it seems doable, but I may be missing some fundamental challenge.

Reply:
To meet certain timelines, can we do these in another PR? (Once we have every part in the Log Viewer put together we will have a better vision of all the requirements as well. )

Reply:
Sg